Fix flaky test in failover2

sungming2 opened the pull request:
### Issue https://github.com/valkey-io/valkey/issues/1640

### Problem

### Solution

### Test

sungming2 commented:
@hpatro @enjoy-binbin could you take a look at this pr please when you have time

enjoy-binbin commented:
> Once the failover rank is upgraded, it should not be downgraded again because it will likely trigger the same epoch election.

can you try to keep this only change and test it again? I did have some doubt when i added this line.

enjoy-binbin commented on src/cluster_legacy.c:
```diff
@@ -3260,17 +3260,18 @@ int clusterProcessPacket(clusterLink *link) {
+            /* Another node has claimed an epoch greater than or equal to ours.
+             * If we have an ongoing election, reset it because we cannot win
+             * with an epoch smaller than or equal to the incoming claim. This
+             * allows us to start a new election as soon as possible. */
```
also please remove these non-related changes.

enjoy-binbin commented on src/cluster_legacy.c:
```diff
@@ -4860,9 +4871,10 @@ void clusterHandleReplicaFailover(void) {
+                                             random() % 100; /* Random delay between 0 and 100 milliseconds. */
         server.cluster->failover_auth_count = 0;
         server.cluster->failover_auth_sent = 0;
+        server.cluster->failover_failed_primary_rank = 0;
```
and this, see line 4886

enjoy-binbin commented on src/cluster_legacy.c:
```diff
@@ -4783,6 +4787,13 @@ void clusterFailoverReplaceYourPrimary(void) {
     /* 5) If there was a manual failover in progress, clear the state. */
     resetManualFailover();
+
+    /* 6) Since we have changed to a new primary node, the previously set
+     * failover_auth_time and failover_auth_sent should no longer be used */
+    if (server.cluster->failover_auth_time) {
+        server.cluster->failover_auth_time = 0;
+        server.cluster->failover_auth_sent = 0;
+    }
```
we have became a new primary, does this affect it? we will no longer enter the failover check logic. you are saying the primary became a replica in a short time and the auth_time is wrong? We do clear the auth_time in clusterSetPrimary

sungming2 replied:
Yes, that was the reason why I added this cleanup. I missed this point `We do clear the auth_time in clusterSetPrimary`.
I will remove it thanks

hpatro commented on src/cluster_legacy.c:
```diff
@@ -4872,7 +4876,7 @@ void clusterHandleReplicaFailover(void) {
          * Specifically 0.5 second * rank. This way those failed primaries will be
          * elected in rank to avoid the vote conflicts. */
```
This comment is no longer true. Could you update it?

hpatro commented on src/cluster_legacy.c:
```diff
@@ -4672,12 +4673,15 @@ int clusterGetFailedPrimaryRank(void) {
 
         /* Skip nodes that do not need to participate in the rank. */
-        if (!nodeFailed(node) || !clusterNodeIsVotingPrimary(node) || node->num_replicas == 0) continue;
+        if (!(nodeFailed(node) || nodeTimedOut(node)) || !clusterNodeIsVotingPrimary(node) || node->num_replicas == 0) continue;
```
It doesn't hurt to explain the conditional statement in code comment. Would be easier to understand why we choose to skip the failed/partially failed node.

hpatro reviewed (commented):
Could you update the high level comment around the behavior of this PR.

sungming2 commented:
> can you try to keep this only change and test it again? I did have some doubt when i added this line.

I tried it hundreds of times, I don't see any specific error due to this change but I can see some same epoch elections because it has insufficient delay

R1
```
162615:S 06 Feb 2025 11:41:23.803 * Start of election delayed for 672 milliseconds (rank #0, primary rank #0, offset 14).
162615:S 06 Feb 2025 11:41:24.210 * Failed primary rank updated to #1, added 500 milliseconds of delay.
162615:S 06 Feb 2025 11:41:25.014 * Starting a failover election for epoch 8, node config epoch is 0, rank is 1
```

R2
```
162730:S 06 Feb 2025 11:41:24.209 * Start of election delayed for 735 milliseconds (rank #0, primary rank #0, offset 14).
162730:S 06 Feb 2025 11:41:25.012 * Starting a failover election for epoch 8, node config epoch is 3, rank is 0
```

enjoy-binbin commented:
I want to know whether the main reason here is delay or because the pfail node is not included in the rank. In the logs i am guessing the main reason is the delay, can this be fixed if we only improve the delay logic? 

From the log, we can see that it is because of the `added 500 milliseconds of delay.` that they initiate the election at the same time. This is a familiar shortcoming and widely accepted. The reason i think is that, the tcl `pause_process` has some delay so these primaries doesn't down in the same time but within 500ms range. Maybe it is a acceptable shortcoming we can try to improve the TCL code? Or we do can improve the delay logic, i haven't figured this out yet (it does happen, no matter how you adjust it, the delay will always randomly conflict)

> Once the failover rank is upgraded, it should not be downgraded again because it will likely trigger the same epoch election.

Is this 3 also caused by the delay design? My original idea was that when a primary is up again, the primaries that were delayed by rank can initiate elections as soon as possible.

sungming2 commented:
I agree with you, but what I am really expecting with this change is it will reduce the chance of elections starting at the same time. (there are cases that pfail node is not included that is why I added this) 

And yes, we can refine this delay logic or implement an entirely new approach later, such as adding a flag to indicate whether a node is in the election process and by comparing the sender's current epoch with its auth epoch, we detect an election conflict and immediately reset it if necessary (could be wrong, just my instant thought).


> My original idea was that when a primary is up again, the primaries that were delayed by rank can initiate elections as soon as possible.

You are right, but I thought it would be better to add a slight delay rather than risk an election conflict since both replicas must wait until the auth timeout (node timeout * 4) in the worst case.

I can just apply increasing delay with small random (1000 + random() %100) if you are ok with it as this is a clear improvement we can do right away. What do you think?

enjoy-binbin commented:
> I can just apply increasing delay with small random only to this pr (1000 + random() %100) if you are ok with it as this is a clear improvement we can do right away. What do you think?

yes, please do it, we also need other eyes to take a look with this delay logic after the changes.

hpatro commented:
@enjoy-binbin what are your thoughts on this? Should we target it for 8.1 ?

enjoy-binbin commented on src/cluster_legacy.c:
```diff
@@ -4675,9 +4676,12 @@ int clusterGetFailedPrimaryRank(void) {
         /* If cluster-replica-validity-factor is enabled, skip the invalid nodes. */
-        if (server.cluster_replica_validity_factor) {
-            if ((now - node->fail_time) > (server.cluster_node_timeout * server.cluster_replica_validity_factor))
+        if (nodeFailed(node) && server.cluster_replica_validity_factor) {
```
do we still need this `nodeFailed(node)`, doesn't the if continue above skip it? Let's also remove the debug log?

sungming2 replied:
I forgot to remove it. Thanks

enjoy-binbin commented:
> what are your thoughts on this? Should we target it for 8.1 ?

We have been running this delay internally for a long time and have not found any problems (or maybe they have not been exposed), so i guess there is no rush. I am ok with this delay logic, but i think this needs an extra eyes. @madolson @PingXie @zuiderkwast

madolson commented:
The update seems OK to me. It feels a bit arbitrary, and it would be nice if it was more dynamic instead of relying on a bunch of internal constants. I'd be okay merging this for 8.1.

zuiderkwast commented:
#2227 only removes the delay if it is known to be safe, in a special case. It falls back to the current behavior.

The changed delays in this PR can still be useful, I think. Sorry, I didn't look at this PR before.
